File: cadwiki-nuget/cadwiki.AutoCAD2021.Base.Utilities/Shared/Blocks.vb
Option Strict On
Option Infer Off
Option Explicit On

Imports System.Runtime.CompilerServices
Imports Autodesk.AutoCAD.ApplicationServices
Imports Autodesk.AutoCAD.DatabaseServices
Imports Autodesk.AutoCAD.EditorInput
Imports Autodesk.AutoCAD.Geometry
Imports Autodesk.AutoCAD.Runtime

Public Class Blocks
    Public Shared Function SortSelectionSetByProximityToPoint(doc As Document, ss As SelectionSet, point As Point3d) _
        As List(Of BlockReference)
        Dim blockRefs As New List(Of BlockReference)
        Dim db As Database = doc.Database
        Using lock As DocumentLock = doc.LockDocument()
            Using t As Transaction = db.TransactionManager.StartTransaction()
                Dim currentSpace As BlockTableRecord = CType(t.GetObject(db.CurrentSpaceId, OpenMode.ForWrite),
                    BlockTableRecord)
                Dim objectIds As ObjectId() = ss.GetObjectIds
                For Each objId As ObjectId In objectIds
                    Dim blockRef As BlockReference = CType(t.GetObject(objId, OpenMode.ForWrite), BlockReference)
                    blockRefs.Add(blockRef)
                Next

                Dim sortedBlockRefs As IEnumerable(Of BlockReference) = From o In blockRefs
                    Order By o.Position.DistanceTo(point)
                Return sortedBlockRefs.ToList()
            End Using
        End Using
        Return Nothing
    End Function

    Public Shared Function GetAttributeValue(doc As Document, blockRef As BlockReference, attributeName As String) _
        As String
        Dim db As Database = doc.Database

        Using lock As DocumentLock = doc.LockDocument()
            Using t As Transaction = db.TransactionManager.StartTransaction()
                For Each attId As ObjectId In blockRef.AttributeCollection
                    Try
                        Dim attRef As AttributeReference = CType(t.GetObject(attId, OpenMode.ForRead, False),
                            AttributeReference)
                        If (attRef.Tag = attributeName) Then
                            Return attRef.TextString
                        End If
                    Catch ex As Exception
                        Debug.WriteLine("Attribute no longer exists because it was deleted.")
                    End Try

                Next
            End Using
        End Using


        Return Nothing
    End Function

    Public Shared Function SetEntityAttributeValue(entity As Entity, attributeName As String, newValue As String) _
        As Boolean
        Dim wasAttributeUpdated As Boolean = False
        Dim doc As Document = Application.DocumentManager.MdiActiveDocument
        Dim objectIdsToInsertionPoints As Dictionary(Of ObjectId, Point3d) = New Dictionary(Of ObjectId, Point3d)
        Dim db As Database = doc.Database
        Using lock As DocumentLock = doc.LockDocument()
            Using transaction As Transaction = db.TransactionManager.StartTransaction()
                Dim blockRefObj As BlockReference = CType(entity, BlockReference)
                Dim attributeCollection As AttributeCollection = blockRefObj.AttributeCollection
                For Each attId As ObjectId In attributeCollection
                    Try
                        Dim obj As DBObject = transaction.GetObject(attId, OpenMode.ForWrite)
                        Dim attribute As AttributeReference = CType(obj, AttributeReference)
                        If (attribute.Tag = attributeName) Then
                            attribute.TextString = newValue
                            wasAttributeUpdated = True
                        End If
                    Catch ex As Exception
                        Debug.WriteLine("Attribute no longer exists because it was deleted.")
                    End Try

                Next
                transaction.Commit()
            End Using
        End Using
        Return wasAttributeUpdated
    End Function

    Public Shared Sub AddAttributesFromBlockTable(doc As Document,
        acBlkTblRec As BlockTableRecord,
        acBlkRef As BlockReference)
        Dim attDefClass As RXClass = RXObject.GetClass(GetType(AttributeDefinition))
        Dim db As Database = doc.Database
        Using lock As DocumentLock = doc.LockDocument
            Using tr As Transaction = doc.TransactionManager.StartTransaction()
                For Each objID As ObjectId In acBlkTblRec
                    If objID.ObjectClass.IsDerivedFrom(attDefClass) Then
                        Dim dbObj As DBObject = tr.GetObject(objID, OpenMode.ForRead)
                        If TypeOf dbObj Is AttributeDefinition Then
                            Dim acAtt As AttributeDefinition = CType(dbObj, AttributeDefinition)
                            If Not acAtt.Constant Then
                                Using acAttRef As New AttributeReference
                                    acAttRef.SetAttributeFromBlock(acAtt, acBlkRef.BlockTransform)
                                    acAttRef.Position = acAtt.Position.TransformBy(acBlkRef.BlockTransform)
                                    acAttRef.TextString = acAtt.TextString
                                    acBlkRef.AttributeCollection.AppendAttribute(acAttRef)
                                    tr.AddNewlyCreatedDBObject(acAttRef, True)
                                End Using
                            End If
                        End If
                    End If
                Next
                tr.Commit()
            End Using
        End Using
    End Sub

    Public Shared Sub DeleteAttributes(doc As Document, acBlkRef As BlockReference)
        Dim attDefClass As RXClass = RXObject.GetClass(GetType(AttributeDefinition))
        Dim db As Database = doc.Database
        Dim purgedIds As ObjectIdCollection = New ObjectIdCollection()

        Using lock As DocumentLock = doc.LockDocument
            Using tr As Transaction = doc.TransactionManager.StartTransaction()
                If acBlkRef.AttributeCollection.Count > 0 Then
                    acBlkRef.UpgradeOpen()
                    For Each attId As ObjectId In acBlkRef.AttributeCollection
                        If attId.IsErased = False Then
                            Dim attRef As AttributeReference = CType(tr.GetObject(attId, OpenMode.ForWrite),
                                AttributeReference)
                            purgedIds.Add(attId)
                            attRef.Erase(True)
                        End If
                    Next
                End If
                db.Purge(purgedIds)
                tr.Commit()
            End Using
        End Using
    End Sub

    Public Shared Function InsertBlockFromFile(doc As Document,
        filePath As String,
        insertionPoint As Point3d,
        rotation As Double,
        layerName As String) As BlockReference
        Dim db As Database = doc.Database
        Using xDb As New Database(False, True)
            xDb.ReadDwgFile(filePath, FileOpenMode.OpenForReadAndReadShare, True, Nothing)
            Using lock As DocumentLock = doc.LockDocument
                Using tr As Transaction = doc.TransactionManager.StartTransaction()
                    Dim name As String = SymbolUtilityServices.GetBlockNameFromInsertPathName(filePath)
                    Dim id As ObjectId = db.Insert(name, xDb, True)
                    If id.IsNull Then
                        doc.Editor.WriteMessage(vbLf & "Failed to insert block")
                        Return Nothing
                    End If
                    Dim currSpace As BlockTableRecord = CType(tr.GetObject(db.CurrentSpaceId, OpenMode.ForWrite),
                        BlockTableRecord)
                    Dim coordS As CoordinateSystem3d = New CoordinateSystem3d(insertionPoint, db.Ucsxdir, db.Ucsydir) _
                    'Determine UCS
                    Dim insert As New BlockReference(insertionPoint, id)
                    insert.Normal = coordS.Zaxis 'Align to UCS
                    insert.Rotation = rotation
                    currSpace.AppendEntity(insert)
                    insert.SetDatabaseDefaults()
                    insert.Layer = layerName
                    tr.AddNewlyCreatedDBObject(insert, True)
                    Dim btr As BlockTableRecord = CType(tr.GetObject(insert.BlockTableRecord, OpenMode.ForWrite),
                        BlockTableRecord)
                    AddAttributesFromBlockTable(doc, btr, insert)
                    'Autodesk.AutoCAD.DatabaseServices.SynchronizeAttributes(btr, insert)
                    tr.Commit()
                    Return insert
                End Using
            End Using

        End Using
    End Function
End Class


Namespace Autodesk.AutoCAD.DatabaseServices
    Module ExtensionMethods
        Public attDefClass As RXClass = RXClass.GetClass(GetType(AttributeDefinition))

        <Extension()>
        Sub SynchronizeAttributes(ByVal target As BlockTableRecord, ByVal instance As BlockReference)
            If target Is Nothing Then Throw New ArgumentNullException("target")
            Dim tr As Transaction = target.Database.TransactionManager.TopTransaction
            If tr Is Nothing Then Throw New Exception(ErrorStatus.NoActiveTransactions)
            Dim attDefs As List(Of AttributeDefinition) = target.GetAttributes(tr)

            For Each id As ObjectId In target.GetBlockReferenceIds(True, False)
                Dim br As BlockReference = CType(tr.GetObject(id, OpenMode.ForWrite), BlockReference)
                br.ResetAttributes(attDefs, tr)
            Next

            If target.IsDynamicBlock = False Then
                target.UpdateAnonymousBlocks()

                For Each id As ObjectId In target.GetAnonymousBlockIds()
                    Dim btr As BlockTableRecord = CType(tr.GetObject(id, OpenMode.ForRead), BlockTableRecord)
                    attDefs = btr.GetAttributes(tr)

                    For Each brId As ObjectId In btr.GetBlockReferenceIds(True, False)
                        Dim br As BlockReference = CType(tr.GetObject(brId, OpenMode.ForWrite), BlockReference)
                        br.ResetAttributes(attDefs, tr)
                    Next
                Next

                attDefs = target.GetAttributes(tr)
                instance.ResetAttributes(attDefs, tr)

            End If
        End Sub

        <Extension()>
        Private Function GetAttributes(ByVal target As BlockTableRecord, ByVal tr As Transaction) _
            As List(Of AttributeDefinition)
            Dim attDefs As List(Of AttributeDefinition) = New List(Of AttributeDefinition)()

            For Each id As ObjectId In target

                If id.ObjectClass = attDefClass Then
                    Dim attDef As AttributeDefinition = CType(tr.GetObject(id, OpenMode.ForRead), AttributeDefinition)
                    attDefs.Add(attDef)
                End If
            Next

            Return attDefs
        End Function

        <Extension()>
        Private Sub ResetAttributes(ByVal br As BlockReference,
            ByVal attDefs As List(Of AttributeDefinition),
            ByVal tr As Transaction)
            Dim attValues As Dictionary(Of String, String) = New Dictionary(Of String, String)()

            For Each id As ObjectId In br.AttributeCollection

                If Not id.IsErased Then
                    Dim attRef As AttributeReference = CType(tr.GetObject(id, OpenMode.ForWrite), AttributeReference)
                    attValues.Add(attRef.Tag,
                        If(attRef.IsMTextAttribute, attRef.MTextAttribute.Contents, attRef.TextString))
                    attRef.[Erase]()
                End If
            Next

            For Each attDef As AttributeDefinition In attDefs
                Dim attRef As AttributeReference = New AttributeReference()
                attRef.SetAttributeFromBlock(attDef, br.BlockTransform)

                If attDef.Constant Then
                    attRef.TextString = If _
                        (attDef.IsMTextAttributeDefinition, attDef.MTextAttributeDefinition.Contents, attDef.TextString)
                ElseIf attValues.ContainsKey(attRef.Tag) Then
                    attRef.TextString = attValues(attRef.Tag)
                End If

                br.AttributeCollection.AppendAttribute(attRef)
                tr.AddNewlyCreatedDBObject(attRef, True)
            Next
        End Sub
    End Module
End Namespace

File: cadwiki-nuget/cadwiki.AC/Utilities/Blocks.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using Autodesk.AutoCAD.ApplicationServices;
using Autodesk.AutoCAD.DatabaseServices;
using Autodesk.AutoCAD.EditorInput;
using Autodesk.AutoCAD.Geometry;
using Autodesk.AutoCAD.Runtime;
using Microsoft.VisualBasic;
using OpenMode = Autodesk.AutoCAD.DatabaseServices.OpenMode;

namespace cadwiki.AC.Utilities
{

    public class Blocks
    {
        public static List<BlockReference> SortSelectionSetByProximityToPoint(Document doc, SelectionSet ss, Point3d point)
        {
            var blockRefs = new List<BlockReference>();
            var db = doc.Database;
            using (var @lock = doc.LockDocument())
            {
                using (var t = db.TransactionManager.StartTransaction())
                {
                    BlockTableRecord currentSpace = (BlockTableRecord)t.GetObject(db.CurrentSpaceId, global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForWrite);
                    ObjectId[] objectIds = ss.GetObjectIds();
                    foreach (ObjectId objId in objectIds)
                    {
                        BlockReference blockRef = (BlockReference)t.GetObject(objId, global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForWrite);
                        blockRefs.Add(blockRef);
                    }

                    IEnumerable<BlockReference> sortedBlockRefs = from o in blockRefs
                                                                  orderby o.Position.DistanceTo(point)
                                                                  select o;
                    return sortedBlockRefs.ToList();
                }
            }
            return null;
        }

        public static string GetAttributeValue(Document doc, BlockReference blockRef, string attributeName)
        {
            var db = doc.Database;

            using (var @lock = doc.LockDocument())
            {
                using (var t = db.TransactionManager.StartTransaction())
                {
                    foreach (ObjectId attId in blockRef.AttributeCollection)
                    {
                        try
                        {
                            AttributeReference attRef = (AttributeReference)t.GetObject(attId, global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForRead, false);
                            if ((attRef.Tag ?? "") == (attributeName ?? ""))
                            {
                                return attRef.TextString;
                            }
                        }
                        catch (global::Autodesk.AutoCAD.Runtime.Exception ex)
                        {
                            Debug.WriteLine("Attribute no longer exists because it was deleted.");
                        }

                    }
                }
            }


            return null;
        }

        public static bool SetEntityAttributeValue(Entity entity, string attributeName, string newValue)
        {
            bool wasAttributeUpdated = false;
            var doc = global::Autodesk.AutoCAD.ApplicationServices.Core.Application.DocumentManager.MdiActiveDocument;
            var objectIdsToInsertionPoints = new Dictionary<ObjectId, Point3d>();
            var db = doc.Database;
            using (var @lock = doc.LockDocument())
            {
                using (var transaction = db.TransactionManager.StartTransaction())
                {
                    BlockReference blockRefObj = (BlockReference)entity;
                    var attributeCollection = blockRefObj.AttributeCollection;
                    foreach (ObjectId attId in attributeCollection)
                    {
                        try
                        {
                            var obj = transaction.GetObject(attId, global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForWrite);
                            AttributeReference attribute = (AttributeReference)obj;
                            if ((attribute.Tag ?? "") == (attributeName ?? ""))
                            {
                                attribute.TextString = newValue;
                                wasAttributeUpdated = true;
                            }
                        }
                        catch (global::Autodesk.AutoCAD.Runtime.Exception ex)
                        {
                            Debug.WriteLine("Attribute no longer exists because it was deleted.");
                        }

                    }
                    transaction.Commit();
                }
            }
            return wasAttributeUpdated;
        }

        public static void AddAttributesFromBlockTable(Document doc, BlockTableRecord acBlkTblRec, BlockReference acBlkRef)
        {
            var attDefClass = RXObject.GetClass(typeof(AttributeDefinition));
            var db = doc.Database;
            using (var @lock = doc.LockDocument())
            {
                using (var tr = doc.TransactionManager.StartTransaction())
                {
                    foreach (ObjectId objID in acBlkTblRec)
                    {
                        if (objID.ObjectClass.IsDerivedFrom(attDefClass))
                        {
                            var dbObj = tr.GetObject(objID, global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForRead);
                            if (dbObj is AttributeDefinition)
                            {
                                AttributeDefinition acAtt = (AttributeDefinition)dbObj;
                                if (!acAtt.Constant)
                                {
                                    using (var acAttRef = new AttributeReference())
                                    {
                                        acAttRef.SetAttributeFromBlock(acAtt, acBlkRef.BlockTransform);
                                        acAttRef.Position = acAtt.Position.TransformBy(acBlkRef.BlockTransform);
                                        acAttRef.TextString = acAtt.TextString;
                                        acBlkRef.AttributeCollection.AppendAttribute(acAttRef);
                                        tr.AddNewlyCreatedDBObject(acAttRef, true);
                                    }
                                }
                            }
                        }
                    }
                    tr.Commit();
                }
            }
        }

        public static void DeleteAttributes(Document doc, BlockReference acBlkRef)
        {
            var attDefClass = RXObject.GetClass(typeof(AttributeDefinition));
            var db = doc.Database;
            var purgedIds = new ObjectIdCollection();

            using (var @lock = doc.LockDocument())
            {
                using (var tr = doc.TransactionManager.StartTransaction())
                {
                    if (acBlkRef.AttributeCollection.Count > 0)
                    {
                        acBlkRef.UpgradeOpen();
                        foreach (ObjectId attId in acBlkRef.AttributeCollection)
                        {
                            if (attId.IsErased == false)
                            {
                                AttributeReference attRef = (AttributeReference)tr.GetObject(attId, global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForWrite);
                                purgedIds.Add(attId);
                                attRef.Erase(true);
                            }
                        }
                    }
                    db.Purge(purgedIds);
                    tr.Commit();
                }
            }
        }

        public static BlockReference InsertBlockFromFile(Document doc, string filePath, Point3d insertionPoint, double rotation, string layerName)
        {
            var db = doc.Database;
            using (var xDb = new Database(false, true))
            {
                xDb.ReadDwgFile(filePath, FileOpenMode.OpenForReadAndReadShare, true, null);
                using (var @lock = doc.LockDocument())
                {
                    using (var tr = doc.TransactionManager.StartTransaction())
                    {
                        string name = SymbolUtilityServices.GetBlockNameFromInsertPathName(filePath);
                        var id = db.Insert(name, xDb, true);
                        if (id.IsNull)
                        {
                            doc.Editor.WriteMessage(Environment.NewLine + "Failed to insert block");
                            return null;
                        }
                        BlockTableRecord currSpace = (BlockTableRecord)tr.GetObject(db.CurrentSpaceId, global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForWrite);
                        var coordS = new CoordinateSystem3d(insertionPoint, db.Ucsxdir, db.Ucsydir);
                        // Determine UCS
                        var insert = new BlockReference(insertionPoint, id);
                        insert.Normal = coordS.Zaxis; // Align to UCS
                        insert.Rotation = rotation;
                        currSpace.AppendEntity(insert);
                        insert.SetDatabaseDefaults();
                        insert.Layer = layerName;
                        tr.AddNewlyCreatedDBObject(insert, true);
                        BlockTableRecord btr = (BlockTableRecord)tr.GetObject(insert.BlockTableRecord, global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForWrite);
                        AddAttributesFromBlockTable(doc, btr, insert);
                        // Autodesk.AutoCAD.DatabaseServices.SynchronizeAttributes(btr, insert)
                        tr.Commit();
                        return insert;
                    }
                }

            }
        }


        public static List<string> GetAllBlockNamesFromDrawing(Document doc)
        {
            List<string> blockNames = new List<string>();
            try
            {
                Database db = doc.Database;
                using (Transaction tr = db.TransactionManager.StartTransaction())
                {
                    BlockTable blockTable = tr.GetObject(db.BlockTableId, OpenMode.ForRead) as BlockTable;
                    if (blockTable != null)
                    {
                        foreach (ObjectId blockId in blockTable)
                        {
                            BlockTableRecord blockRecord = tr.GetObject(blockId, OpenMode.ForRead) as BlockTableRecord;
                            if (blockRecord != null)
                            {
                                blockNames.Add(blockRecord.Name);
                            }
                        }
                    }
                    tr.Commit();
                }
            }
            catch (System.Exception ex)
            {
                ExceptionHandler.WriteToEditor(ex);
            }

            return blockNames;
        }
    }


    namespace Autodesk.AutoCAD.DatabaseServices
    {
        static class ExtensionMethods
        {
            public static RXClass attDefClass = RXObject.GetClass(typeof(AttributeDefinition));

            public static void SynchronizeAttributes(this BlockTableRecord target, BlockReference instance)
            {
                if (target is null)
                    throw new ArgumentNullException("target");
                var tr = target.Database.TransactionManager.TopTransaction;
                if (tr is null)
                    throw new global::Autodesk.AutoCAD.Runtime.Exception(ErrorStatus.NoActiveTransactions);
                var attDefs = target.GetAttributes(tr);

                foreach (ObjectId id in target.GetBlockReferenceIds(true, false))
                {
                    BlockReference br = (BlockReference)tr.GetObject(id, global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForWrite);
                    br.ResetAttributes(attDefs, tr);
                }

                if (target.IsDynamicBlock == false)
                {
                    target.UpdateAnonymousBlocks();

                    foreach (ObjectId id in target.GetAnonymousBlockIds())
                    {
                        BlockTableRecord btr = (BlockTableRecord)tr.GetObject(id, global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForRead);
                        attDefs = btr.GetAttributes(tr);

                        foreach (ObjectId brId in btr.GetBlockReferenceIds(true, false))
                        {
                            BlockReference br = (BlockReference)tr.GetObject(brId, global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForWrite);
                            br.ResetAttributes(attDefs, tr);
                        }
                    }

                    attDefs = target.GetAttributes(tr);
                    instance.ResetAttributes(attDefs, tr);

                }
            }

            private static List<AttributeDefinition> GetAttributes(this BlockTableRecord target, Transaction tr)
            {
                var attDefs = new List<AttributeDefinition>();

                foreach (ObjectId id in target)
                {

                    if (id.ObjectClass == attDefClass)
                    {
                        AttributeDefinition attDef = (AttributeDefinition)tr.GetObject(id, global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForRead);
                        attDefs.Add(attDef);
                    }
                }

                return attDefs;
            }

            private static void ResetAttributes(this BlockReference br, List<AttributeDefinition> attDefs, Transaction tr)
            {
                var attValues = new Dictionary<string, string>();

                foreach (ObjectId id in br.AttributeCollection)
                {

                    if (!id.IsErased)
                    {
                        AttributeReference attRef = (AttributeReference)tr.GetObject(id, global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForWrite);
                        attValues.Add(attRef.Tag, attRef.IsMTextAttribute ? attRef.MTextAttribute.Contents : attRef.TextString);
                        attRef.Erase();
                    }
                }

                foreach (AttributeDefinition attDef in attDefs)
                {
                    var attRef = new AttributeReference();
                    attRef.SetAttributeFromBlock(attDef, br.BlockTransform);

                    if (attDef.Constant)
                    {
                        attRef.TextString = attDef.IsMTextAttributeDefinition ? attDef.MTextAttributeDefinition.Contents : attDef.TextString;
                    }
                    else if (attValues.ContainsKey(attRef.Tag))
                    {
                        attRef.TextString = attValues[attRef.Tag];
                    }

                    br.AttributeCollection.AppendAttribute(attRef);
                    tr.AddNewlyCreatedDBObject(attRef, true);
                }
            }
        }
    }
}